### PR TITLE
drivers: sensor: bq274xx: Configure or confirm chemistry profile

### PR DESCRIPTION
**Pull request by topisani**
Both the BQ27421 and BQ27427 have a few preset Chemistry profiles.
For the BQ27421 there exists three variants of the IC, and for the BQ27427,
it can be configured. The chemistry profile among other things includes the
taper voltage, which is used to detect charge termination.

This adds an optional `chemistry-id` config option to the driver. On the
BQ27421, it will confirm that the correct variant of the IC is mounted,
and on the BQ27427, it will configure it with the correct value.

Side note: The reference manual for the BQ27427
(https://www.ti.com/lit/ug/sluucd5/sluucd5.pdf) currently contains some
errors and inconsistencies regarding these registers. The table on page 7
appears to be correct.

### CODE REVIEW

File: drivers/sensor/bq274xx/bq274xx.h [hunk 1]
```diff
@@ -60,15 +60,15 @@
...
-#define BQ274XX_EXT_BLKDAT(off)                (BQ274XX_EXT_BLKDAT_START + off)
+#define BQ274XX_EXT_OPCONFIG     0x3A /* OpConfig() */
+#define BQ274XX_EXT_CAPACITY     0x3C /* DesignCapacity() */
+#define BQ274XX_EXT_DATA_CLASS   0x3E /* DataClass() */
+#define BQ274XX_EXT_DATA_BLOCK   0x3F /* DataBlock() */
+#define BQ274XX_EXT_BLKDAT_START 0x40 /* BlockData_start() */
+#define BQ274XX_EXT_BLKDAT_END   0x5F /* BlockData_end() */
+#define BQ274XX_EXT_CHECKSUM     0x60 /* BlockDataCheckSum() */
+#define BQ274XX_EXT_DATA_CONTROL 0x61 /* BlockDataControl() */
+#define BQ274XX_EXT_BLKDAT(off)  (BQ274XX_EXT_BLKDAT_START + off)
```
Comment (MaureenHelm) [inline drivers/sensor/bq274xx/bq274xx.h:71]: There are a lot of whitespace changes unrelated to the chemistry changes. Please revert the whitespace changes or move them to a separate commit

File: dts/bindings/sensor/ti,bq274xx.yaml [hunk 2]
```diff
@@ -31,6 +31,13 @@ properties:
     required: true
     description: Battery Terminate Voltage in mV
 
+  chemistry-id:
+    type: int
+    default: 0
```
Comment (MaureenHelm) [inline dts/bindings/sensor/ti,bq274xx.yaml:36]: The description needs to provide a justification for the default. Is it a register reset value, for example?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/bq274xx/bq274xx.h (hunk 1 above)
Comment (fabiobaltieri) [reply drivers/sensor/bq274xx/bq274xx.h:71]: Or not do them at all, I could understand if you were about to do a major driver rework and maintain it but just for the sake of it does not help much. I understand you probably ran clang-format or something but then other contributors don't use it, it'll just go back out of sync again anyway.

Comment (topisani) [reply drivers/sensor/bq274xx/bq274xx.h]: Fair enough, i reverted them. I ran clang-format to ensure i was following the style guide, but i understand it made the diff harder to read.

File: dts/bindings/sensor/ti,bq274xx.yaml (hunk 2 above)
Comment (topisani) [reply dts/bindings/sensor/ti,bq274xx.yaml:36]: I slightly changed the description, I'm not entirely sure it helps. I needed a sentinel value, as you may not want to check the register in case you want to just assume the correct IC variant of the BQ27421 has been selected to match the battery on the board.

I chose zero, as it is not used as a chemistry ID. I could go with -1 to be fully outside the range of possible values.

**Review by MaureenHelm — DISMISSED** [review]

File: drivers/sensor/bq274xx/bq274xx.c [hunk 3]
```diff
@@ -321,6 +321,62 @@ static int bq27427_ccgain_quirk(const struct device *dev)
...
+		switch (val) {
+		case 0x3230:
+			cmd = 0x0030;
+			break;
+		case 0x1202:
+			cmd = 0x0031;
+			break;
+		case 0x3142:
+			cmd = 0x0032;
+			break;
```
Comment (fabiobaltieri) [inline drivers/sensor/bq274xx/bq274xx.c:365]: this can use a bunch of `#define`, how about:

- add defines for the cmd values (`BQ274XX_CHEM_A_CMD`... or something like that, in `bq274xx.h`)
- add a set of defines that can be used in the device tree for the value, those would have to go in a new `dt-bindings/sensor/bq274xx.h`, see the other files there

**Review by fabiobaltieri — CHANGES_REQUESTED** [review]

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
Few more compliance warnings, btw you can run that locally with `./scripts/ci/check_compliance.py`

**Comment by topisani** [comment]
> Few more compliance warnings, btw you can run that locally with `./scripts/ci/check_compliance.py`

Thanks, that helps a lot!

**Review by fabiobaltieri — DISMISSED** [review]

File: dts/bindings/sensor/ti,bq274xx.yaml (hunk 2 above)
Comment (aasinclair) [reply dts/bindings/sensor/ti,bq274xx.yaml:36]: I suggest removing the default from here, and adding it in BQ274XX_INIT.

File: drivers/sensor/bq274xx/bq274xx.c [hunk 4]
```diff
@@ -798,6 +866,7 @@ static const struct sensor_driver_api bq274xx_battery_driver_api = {
 		.design_capacity = DT_INST_PROP(index, design_capacity),	\
 		.taper_current = DT_INST_PROP(index, taper_current),		\
 		.terminate_voltage = DT_INST_PROP(index, terminate_voltage),	\
+		.chemistry_id = DT_INST_PROP(index, chemistry_id),			\
```
Comment (aasinclair) [inline drivers/sensor/bq274xx/bq274xx.c:869]:
```suggestion
		.chemistry_id = DT_INST_PROP_OR(index, chemistry_id, 0),			\
```

File: drivers/sensor/bq274xx/bq274xx.c [hunk 5]
```diff
@@ -321,6 +321,69 @@ static int bq27427_ccgain_quirk(const struct device *dev)
...
+			break;
+		case BQ27427_CHEM_ID_B:
+			cmd = BQ27427_CTRL_CHEM_B;
+			break;
+		case BQ27427_CHEM_ID_C:
+			cmd = BQ27427_CTRL_CHEM_C;
+			break;
+		default:
+			LOG_ERR("Unsupported chemistry ID 0x%04x", val);
+			return -EIO;
```
Comment (aasinclair) [inline drivers/sensor/bq274xx/bq274xx.c:375]: EIO is used for I2C failure, so I suggest using something else if the I2C is OK, maybe:
```suggestion
			return -EINVAL;
```

**Review by aasinclair — CHANGES_REQUESTED** [review]

File: drivers/sensor/bq274xx/bq274xx.c (hunk 4 above)
Comment (topisani) [reply drivers/sensor/bq274xx/bq274xx.c]: done

File: drivers/sensor/bq274xx/bq274xx.c (hunk 5 above)
Comment (topisani) [reply drivers/sensor/bq274xx/bq274xx.c]: done

**Review by fabiobaltieri — DISMISSED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Review by aasinclair — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]
